invokeai/backend/model_manager/download/queue.py

```diff
@@ -0,0 +1,689 @@
+
+    def start_all_jobs(self):
+        """Start (enqueue) all jobs that are idle or paused."""
+        try:
+            self._lock.acquire()
+            for job in self._jobs.values():
+                if job.status in [DownloadJobStatus.IDLE, DownloadJobStatus.PAUSED]:
+                    self.start_job(job)
+        finally:
+            self._lock.release()
+
+    def pause_all_jobs(self):
+        """Pause all running jobs."""
+        try:
+            self._lock.acquire()
+            for job in self._jobs.values():
+                if job.status == DownloadJobStatus.RUNNING:
+                    self.pause_job(job)
+        finally:
+            self._lock.release()
+
+    def cancel_all_jobs(self, preserve_partial: bool = False):
+        """Cancel all running jobs."""
+        try:
+            self._lock.acquire()
+            for job in self._jobs.values():
+                if not self._in_terminal_state(job):
+                    self.cancel_job(job, preserve_partial)
+        finally:
+            self._lock.release()
+
+    def _in_terminal_state(self, job: DownloadJobBase):
+        return job.status in [
+            DownloadJobStatus.COMPLETED,
+            DownloadJobStatus.ERROR,
+            DownloadJobStatus.CANCELLED,
+        ]
+
+    def _start_workers(self, max_workers: int):
+        """Start the requested number of worker threads."""
+        for i in range(0, max_workers):
+            worker = threading.Thread(target=self._download_next_item, daemon=True)
+            worker.start()
+            self._worker_pool.add(worker)
+
+    def _download_next_item(self):
+        """Worker thread gets next job on priority queue."""
+        done = False
+        while not done:
+            job = self._queue.get()
+
+            try:  # this is for debugging priority
+                self._lock.acquire()
+                job.job_sequence = self._sequence
+                self._sequence += 1
+            finally:
+                self._lock.release()
+
+            if job == STOP_JOB:  # marker that queue is done
+                done = True
+
+            if job.status == DownloadJobStatus.ENQUEUED:  # Don't do anything for non-enqueued jobs (shouldn't happen)
+                # There should be a better way to dispatch on the job type
+                if not self._quiet:
+                    self._logger.info(f"{job.source}: Downloading to {job.destination}")
+                if isinstance(job, DownloadJobURL):
+                    self._download_with_resume(job)
+                elif isinstance(job, DownloadJobRepoID):
+                    self._download_repoid(job)
+                elif isinstance(job, DownloadJobPath):
+                    self._download_path(job)
+                else:
+                    raise NotImplementedError(f"Don't know what to do with this job: {job}, type={type(job)}")
+
+            if job.status == DownloadJobStatus.CANCELLED:
+                self._cleanup_cancelled_job(job)
```

Comment:
It's intentional. We want to be able to detect when the job's status changes to CANCELLED or PAUSED so that we can take action. This is likely to happen during the `_download_*()` method and locking will prevent the change from taking effect during the lifetime of the job.